Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,472 @@
+
+#include "msblenUtils.h"
+#include "MeshSync/SceneGraph/msTexture.h"
+#include "MeshSync/Utility/msMaterialExt.h"
+
+namespace blender {
+// Blender uses hardcoded string identifiers to figure out what the sockets do:
+const auto baseColorIdentifier = "Base Color";
+const auto colorIdentifier = "Color";
+const auto roughnessIdentifier = "Roughness";
+const auto metallicIdentifier = "Metallic";
+const auto normalIdentifier = "Normal";
+const auto normalStrengthIdentifier = "Strength";
+const auto surfaceIdentifier = "Surface";
+const auto emissionIdentifier = "Emission";
+const auto emissionStrengthIdentifier = "Emission Strength";
+
+const auto shaderIdentifier = "Shader";
+
+const auto displacementIdentifier = "Displacement";
+const auto heightIdentifier = "Height";
+const auto scaleIdentifier = "Scale";
+
+bNode* removeReroutes(bNode* node, const Material* mat) {
+	if (!node) {
+		return nullptr;
+	}
+
+	auto tree = mat->nodetree;
+	if (node->type == NODE_REROUTE)
+	{
+		for (auto link : list_range((bNodeLink*)tree->links.first)) {
+			if (link->tonode == node) {
+				return removeReroutes(link->fromnode, mat);
+			}
+		}
+
+		return nullptr;
+	}
+
+	return node;
+}
+
+bNode* getNodeConnectedToSocket(bNodeSocket* socket) {
+	if (socket->link) {
+		return socket->link->fromnode;
+	}
+
+	return nullptr;
+}
+
+bool getBSDFAndOutput(const Material* mat, bNode*& bsdf, bNode*& output) {
+	// Find BSDF connected to an output node:
+	auto tree = mat->nodetree;
+	for (auto link : list_range((bNodeLink*)tree->links.first)) {
+		if (link->tonode &&
+			link->tonode->type == SH_NODE_OUTPUT_MATERIAL &&
+			STREQ(link->tosock->identifier, surfaceIdentifier)) {
+			bsdf = removeReroutes(link->fromnode, mat);
+
+			// If there is a mix shader, use the first input we can find that has a connection:
+			if (bsdf && bsdf->type == SH_NODE_MIX_SHADER) {
+				for (auto inputSocket : list_range((bNodeSocket*)bsdf->inputs.first)) {
+					if (STREQ(inputSocket->name, shaderIdentifier)) {
+						bNode* connectedBSDF = removeReroutes(getNodeConnectedToSocket(inputSocket), mat);
+						if (connectedBSDF)
+						{
+							bsdf = connectedBSDF;
+							break;
+						}
+					}
+				}
+			}
+
+			output = link->tonode;
+			return bsdf && output;
+		}
+	}
+
+	return false;
+}
+
+bNodeSocket* getInputSocket(bNode* node, const char* socketName) {
+	for (auto inputSocket : list_range((bNodeSocket*)node->inputs.first)) {
+		if (STREQ(inputSocket->name, socketName)) {
+			return inputSocket;
+		}
+	}
+
+	return nullptr;
+}
+
+int msblenMaterialsExportHelper::exportTexture(const std::string& path, ms::TextureType type) const
+{
+	return m_texture_manager->addFile(path, type);
+}
+
+void msblenMaterialsExportHelper::handleImageNode(ms::TextureType& textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode)
+{
+	if (sourceNode->id) {
```

Comment:
I've created a method. Not sure how useful it is to have so many methods for every little piece of functionality that's only used once though. It can be harder to read than having it all in a single place.